ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/SelectionUtil.java

```diff
@@ -65,13 +71,27 @@ public static LogicalSchema buildProjectionSchema(
           && parentSchema.isKeyColumn(((ColumnReferenceExp) expression).getColumnName());
 
       if (keyColumn) {
-        builder.keyColumn(select.getAlias(), expressionType);
-        keys.put(select.getAlias(), expressionType);
+        sourceKeyToAliasAndType.put(
+            ((ColumnReferenceExp) expression).getColumnName(),
+            new Pair<>(select.getAlias(), expressionType)
+        );
       } else {
         builder.valueColumn(select.getAlias(), expressionType);
       }
     }
 
+    // keys need to be added in the same order as the source
+    for (final Column key : parentSchema.key()) {
+      final List<Pair<ColumnName, SqlType>> aliasAndTypes = sourceKeyToAliasAndType.get(key.name());
+      for (final Pair<ColumnName, SqlType> aliasAndType : aliasAndTypes) {
+        if (aliasAndType != null) {
+          // can be null if the key was not selected - this will cause a failure
```

Comment:
Do you refer to `aliasAndType` -- I don't see how it could be `null`? If could be that the `Pair` elements are `null` though?

I think we can remove the ` if (aliasAndType != null) {` check?

Reply:
big oops! I did a last minute change here and meant to have the null check on `aliasAndTypes`. It should fail the tests with an NPE

Reply:
actually, now that i've changed to `MultiMap` it'll never return `null`, just an empty list :) so we're all good to remove the null check